runtime/src/iree/hal/utils/caching_allocator.h

```diff
@@ -0,0 +1,98 @@
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// A HAL buffer allocator that caches allocations instead of returning them to
+// the underlying device allocator.
+//
+// Allocation limits can be independently tuned per heap they originate from to
+// enable heavier caching of more expensive allocations such as mappable
+// device-local and host-visible buffers on devices with discrete memory.
+// Pools are scanned in-order to allow for prioritization.
+//
+// Thread-safe: the allocator can be shared across multiple user-level devices
+// manipulated from multiple threads.
+typedef struct iree_hal_caching_allocator_t iree_hal_caching_allocator_t;
+
+// Parameters used to configure an iree_hal_caching_allocator_t pool.
+// These cannot be changed once the allocator has been created.
+typedef struct iree_hal_caching_allocator_pool_params_t {
+  // Underlying allocator heap that services allocation requests for the pool.
+  //
+  // Additional flags may be added on top of what the underlying heap supports
+  // such as IREE_HAL_MEMORY_TYPE_TRANSIENT to limit a pool to only working with
+  // transient buffers.
+  iree_hal_allocator_memory_heap_t heap;
+
+  // Maximum size of an allocation in bytes; larger allocations will be sent
+  // directly through to the underlying allocator.
+  iree_device_size_t max_allocation_size;
+
+  // Maximum total size of all allocations made from the pool that will be
+  // retained. After this limit is reached allocation requests will be sent
+  // directly through to the underlying allocator.
+  iree_device_size_t max_allocation_capacity;
+
+  // Maximum number of free allocations that will be tracked.
+  // This is used to allocate storage for the free list and should be reasonably
+  // bounded (~64-1024).
+  iree_host_size_t max_free_allocation_count;
+} iree_hal_caching_allocator_pool_params_t;
+
+// Initializes |out_params| to the default values using |heap| for storage.
+void iree_hal_caching_allocator_pool_params_initialize(
+    iree_hal_allocator_memory_heap_t heap,
+    iree_hal_caching_allocator_pool_params_t* out_params);
+
+// Creates an allocator that caches allocations using |device_allocator| for
+// serving requests.
+//
+// All allocations will be cached until the allocator is trimmed and in highly
+// dynamic programs this can easily exceed available memory. Prefer using
+// explicit pools per heap with maximum sizes for safer behavior and better
+// tuning (limit caching to expensive heaps, etc).
+//
+// Buffer import and export and other operations that the caching allocator
+// cannot service will be directed to the underlying |device_allocator|.
+//
+// Thread-safe: internal synchronization of caching allocator data structures
+// allows multiple threads to allocate and free buffers.
+iree_status_t iree_hal_caching_allocator_create_unbounded(
+    iree_hal_allocator_t* device_allocator, iree_allocator_t host_allocator,
+    iree_hal_allocator_t** out_allocator);
```

Comment:
Would unit tests be useful for this? Most of the caching is an implementation detail, but we could at least test that the wrapped allocator still implements the base allocator API correctly.

Reply:
for sure - I'll take a peek at how messy that would be (need to mock out the base allocator, which entails effectively writing an allocator, and that scares me) - the CTS would be ideal for testing things like this as an intermediate between unit testing and e2e (and would be useful for any custom allocator)

Reply:
https://github.com/iree-org/iree/blob/main/runtime/src/iree/hal/cts/allocator_test.h is pretty limited in what it covers today (partly because the API is pretty broad and implementations can choose what they want to / what they can actually support)

It would be neat to test the underlying allocators and then the those allocators wrapped in others

Reply:
Yeah, the CTS is limited ;( may need to rework it at some point. We should be able to have a large suite for all expected usage by the compiler by querying for compatibility like the compiler does and conditionally skipping.